### DISCUSSION TITLE
Issue in running the Standalone nek.i tutorial input file

### DISCUSSION THREAD

**Discussion by AnthonyB08** [neams-th-coe/cardinal#600] [Q&A]
Hi all, 

I am now beginning my journey of cardinal simulation with my first task, simulating a simple turbulent N-S problem. To accomplish this I have created my own directory in cardinal called "Projects" and within that directory "NS". I copied all of the files from the standalone tutorial into the NS directory as such: 
cp -R /home/anthonyb08/cardinal/tutorials/standalone/* /home/anthonyb08/cardinal/Projects/NS

I then cd into the NS directory to run the simulation using the following command: 
 mpiexec -np 2 ../../cardinal-opt -i nek.i --n-threads=2

The simulation begins to run up until this point :

 verify mesh topology
 -0.50000000000000000       0.50000000000000000       Xrange
 -0.50000000000000000       0.50000000000000000       Yrange
   0.0000000000000000        20.000000000000004       Zrange
 done :: verify mesh topology

 mesh metrics:
 GLL grid spacing min/max    : 6.31E-04 6.98E-02
 scaled Jacobian  min/max/avg: 3.68E-01 9.35E-01 7.98E-01
 aspect ratio     min/max/avg: 2.69E+00 3.39E+01 1.14E+01

 call usrdat3
 done :: usrdat3

gridpoints unique/tot:       2263717      3317760
dofs vel/pr:                 2187780      2258340
 nek setup done in    2.5278E+00 s

 set initial conditions
 nekuic (1) for ifld            1
 call nekuic for vel  
 xyz min   -0.50000     -0.50000       0.0000    
 uvwpt min  0.45434E-10 -0.99997E-01 -0.21000       0.0000       0.0000
used config: pw+host (MPI min/max/avg: 4.50e-06s 1.22e-05s 8.82e-06s / avg bi-bw: 14.6GB/s/rank)
min  0% of the local elements are internal

meshParallelGatherScatterSetup N=7
..timing gs: 8.42e-03s 9.65e-03s 9.98e-03s 9.86e-03s
used config: pw+host (MPI min/max/avg: 9.60e-06s 1.94e-05s 1.45e-05s / avg bi-bw: 26.7GB/s/rank)
copying solution from nek
calling udf_setup ... .--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 1 with PID 0 on node DESKTOP-LTJ793R exited on signal 9 (Killed).
--------------------------------------------------------------------------

What does the community suggest?

-Thanks AB

**Comment by aprilnovak** [comment]
Hmm, I am not sure. This tutorial is part of our test suite, so we know it runs fine and is up-to-date.

Can you try re-running without threads? Nek only uses MPI. 

```
rm -rf .cache/
mpiexec -np 8 cardinal-opt -i nek.i
```

That first line is deleting the JIT (just-in-time) compiled files that Nek creates when running an input file. Normally you don't need to do anything special with them, but it's good practice if input files are being troublesome to delete those so you're 100% sure there's not any lingering build settings/etc. from previous runs.

**Reply by AnthonyB08** [reply]
Strange, I get the same error. My installation passed all of the checks, run OpenMC and Neks separately, as well as the Cardinal test suites. I'll keep looking into the issue.

**Reply by aprilnovak** [reply]
The test suite could be skipping this particular test, because we currently require at least 8 ranks to run it. 

I'd suggest proceeding to the other tutorials/examples, and I'll focus on fixing whatever is wrong with this one. If the whole test suite passes, then your install is fine and it's probably some issue on our end (like the case is using too much memory, or something we should fix).

**Reply by AnthonyB08** [reply]
Okay, I went to the standalone channel simulation in the test directory (ie test/tests/nek_standalone/channel/nek.i ) and was able to run the simulation successfully. I will work with these input files for now. Thank you very much for your help!

**Comment by ME-20244** [comment]
Hi All !
I have a problem in running nekrs within cardinal. 
I looked to the majority of cardinal discussions related to my problem, but no success till now.
Here are the main steps I used to resolve the problem:
1- I installed and run nekrs as a stand alone program in CPU mode: OK!
2- I installed cardinal without conda (as indicated to include nekrs). 
3- Some of tutorials were tested with openMC without errors, even for advanced cases on coupling.
4- When I tried to include nekrs, the execution fails:
4-1/ Failure when I lunch nekrs example even as a standalone case within cardinal. 
4-2/ When I lunch unit test as indicated in steps of installation ; it is OK !
4-3/ When I lunch ./run_tests -j8, I get :
* errors of OpenMC's: Exodiff/CSVDiff ; even cross section data is available and the path is correctly defined. 
  example of output:test:neutronics/flux.flux ................................................................... FAILED (CSVDIFF)

* errors of nekrs, mainly  :
--------------------------------------------------------------------------------------------------------------
test:nek_errors/invalid_settings.executioner ................................. FAILED (EXPECTED ERROR MISSING)
test:userobjects/subchannel_layered.type_error ............................... FAILED (EXPECTED ERROR MISSING)
test:nek_errors/insufficient_scratch.too_small_a ............................. FAILED (EXPECTED ERROR MISSING)
test:nek_errors/invalid_scalar.invalid_field_scalar01 ........................ FAILED (EXPECTED ERROR MISSING)
test:conduction/zero_flux.zero_flux_total ..................................................... FAILED (CRASH)
test:nek_separatedomain/invalid_params.invalid_inlet_boundary_multiple ....... FAILED (EXPECTED ERROR MISSING)
test:nek_errors/invalid_field.invalid_field .................................. FAILED (EXPECTED ERROR MISSING)
test:userobjects/layered_layered.multiple_layers .............................................. FAILED (CRASH)
test:userobjects/hexagonal_gap_layered.type_error ............................ FAILED (EXPECTED ERROR MISSING)
test:nek_stochastic.driver_transient .......................................................... FAILED (CRASH)
test:nek_errors/deformation.missing_mesh_solver .............................. FAILED (EXPECTED ERROR MISSING)
test:userobjects/sideset_layered.invalid_boundary ............................ FAILED (EXPECTED ERROR MISSING)
test:multiple_nek_apps/two_channels.invalid_write_fld ........................ FAILED (EXPECTED ERROR MISSING)
test:nek_stochastic/errors.exceed_available .................................. FAILED (EXPECTED ERROR MISSING)
test:nek_errors/deformation/user.displacements ............................... FAILED (EXPECTED ERROR MISSING)
test:nek_separatedomain/transfers_temperature.inlet_t ......................................... FAILED (CRASH)
test:postprocessors/nek_pressure_surface_force.pressure ....................................... FAILED (CRASH)
test:nek_output.scalar_output ................................................................. FAILED (CRASH)
test:conduction/boundary_and_volume/prism.duplicate_temp ..................... FAILED (EXPECTED ERROR MISSING)
test:nek_mesh/exact.boundary_exact ............................................................ FAILED (CRASH)
test:userobjects/hexagonal_gap_layered/normals.normal_component ............................... FAILED (CRASH)
test:nek_errors/insufficient_scratch/standalone.ok_standalone ................................. FAILED (CRASH)
test:userobjects/radial_layered.1d_output ..................................................... FAILED (CRASH)
test:userobjects/interval.interval ............................................................ FAILED (CRASH)
test:nek_temp/exact.exact_temperature ......................................................... FAILED (CRASH)
test:userobjects/nek_scalar_value.controls .................................................... FAILED (CRASH)
test:nek_temp/second_order.second_order_temperature ........................................... FAILED (CRASH)
test:nek_temp/first_order.first_order_temperature ............................................. FAILED (CRASH)
test:nek_standalone/start_time.wrapper ........................................................ FAILED (CRASH)
--------------------------------------------------------------------------------------------------------------
Please, any hints to surmount this issue.

**Reply by aprilnovak** [reply]
Hi @ME-20244, can you try running steps 1 and 3 from the "Checking the Install" section here: https://cardinal.cels.anl.gov/without_conda.html

and then please report whether those cases ran here? Can you also try re-running the test suite, but save the screen output like `./run_tests > out.txt` and then attach `out.txt` here? I'll need to see the entire output to diagnose why the tests failed.

**Comment by ME-20244** [comment]
Hi!
Thank you very much for your prompt reply and your valuable help.
[Test_Results.txt](https://github.com/neams-th-coe/cardinal/files/15232541/Test_Results.txt)

Yes, the three first tests run without issues (see below). For the whole tests, I attached the output file.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
TEST-1    OK
============================>     RESULTS     <============================

 k-effective (Collision)     = 0.94693 +/- 0.02293
 k-effective (Track-length)  = 0.95170 +/- 0.03202
 k-effective (Absorption)    = 0.92054 +/- 0.01361
 Leakage Fraction            = 0.37067 +/- 0.00504

openmc0: Extracting OpenMC tallies... done
openmc0:  Solve Converged!
openmc0: 
openmc0: Postprocessor Values:
openmc0: +----------------+-------------------+----------------+-------------------+
openmc0: | time           | fluid_heat_source | heat_source    | solid_heat_source |
openmc0: +----------------+-------------------+----------------+-------------------+
openmc0: |   0.000000e+00 |      0.000000e+00 |   0.000000e+00 |      0.000000e+00 |
openmc0: |   1.000000e+00 |      0.000000e+00 |   5.000000e+02 |      5.000000e+02 |
openmc0: |   2.000000e+00 |      0.000000e+00 |   5.000000e+02 |      5.000000e+02 |
openmc0: +----------------+-------------------+----------------+-------------------+
openmc0: 

TEST-2   OK
warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
/home/me/.local/lib/python3.10/site-packages/openmc/model/funcs.py:114: FutureWarning: The rectangular_prism(...) function has been replaced by the RectangularPrism(...) class. Future versions of OpenMC will not accept rectangular_prism.
  warn("The rectangular_prism(...) function has been replaced by the "
/home/me/.local/lib/python3.10/site-packages/openmc/source.py:539: FutureWarning: This class is deprecated in favor of 'IndependentSource'
  warnings.warn("This class is deprecated in favor of 'IndependentSource'", FutureWarning)

TEST-3
>>> runtime statistics (step= 40  totalElapsed= 112.621s):
name                    time          abs%  rel%  calls
  solve                 4.03082e+01s  100.0      
    min                 2.91107e-01s
    max                 3.35604e+00s
    flops/rank          5.47939e+09
    checkpointing       1.23614e-02s   0.0        1
    udfExecuteStep      5.50140e-05s   0.0        40
    makef               4.21418e-01s   1.0        40
    makeq               1.45221e-01s   0.4        40
    udfProperties       2.23731e-04s   0.0        41
    velocitySolve       2.86971e+00s   7.1        40
      rhs               2.33270e-01s   0.6   8.1  40
    pressureSolve       3.19579e+01s  79.3        40
      rhs               4.51741e-01s   1.1   1.4  40
      preconditioner    2.48366e+01s  61.6  77.7  1312
        pMG smoother    1.82456e+01s  45.3  73.5  2624
        pMG smoother    4.83074e+00s  12.0  19.5  2624
        coarse grid     1.91070e-01s   0.5   0.8  1312
      initial guess     2.39136e-01s   0.6   0.7  40
    scalarSolve         4.37724e+00s  10.9        40
      rhs               1.79053e-02s   0.0   0.4  40

finished with exit code 0
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

**Reply by aprilnovak** [reply]
Hi @ME-20244 - it looks like you're running into issues on both the OpenMC and NekRS sides.

### OpenMC errors

Your OpenMC errors look like small exo/csv diffs (i.e. the code is running fine, but there's small differences in the results). What OS are you running on? The test suite should run fine on a Linux-based OS, but it's known that OpenMC's test suite _runs_ but does not give 100% matches for the simulation outputs on MacOS (https://github.com/openmc-dev/openmc/issues/801). In practice, this isn't something to be that concerned about because these differences in output get smaller as you increase the number of particles (the differences should be within statistical error).

### Nek errors

These look like a NekRS-specific error, not something in Cardinal itself. Can you try running NekRS by itself, without Cardinal? To do so, you'll run

```
cardinal/install/bin/nrsmpi <casename> <mpi ranks>
```

Can you post what the output of that is here? For instance, to run the `sfr_7pin` tutorial on 2 MPI ranks and save the console output to a file, this command is

```
cardinal/install/bin/nrsmpi sfr_7pin 2 > out.txt
```

**Comment by ME-20244** [comment]
Hi!
For the operating system, I'm using Linux ubuntu22.04.
When running the test as suggested, I get the error message: CUDA Error [ 101 ]: CUDA_ERROR_INVALID_DEVICE;
To run it, I forced the backend to serial. 
I attached the output here, and it seems that I ran out of memory!
[out.txt](https://github.com/neams-th-coe/cardinal/files/15249909/out.txt)

**Reply by aprilnovak** [reply]
Oops, I accidentally told you to run the wrong case (I needed to tell you to run a NekRS standalone case). Can you try the following instead? 

```
cd tutorials/standalone
cardinal/install/bin/nrsmpi turbPipe 2 2>&1 out.txt
```

**Comment by ME-20244** [comment]
Here it is !
[out.txt](https://github.com/neams-th-coe/cardinal/files/15250386/out.txt)
Thanks in advance.

**Reply by aprilnovak** [reply]
Ok that looks normal - can you also please try the following? This is a test which is failing in your run of the test suite, so I want to see if it runs okay outside the test harness.

```
cd cardinal/test/tests/conduction/zero_flux
mpiexec -np 2 cardinal-opt -i main.i 2>&1 out.txt
```

**Comment by ME-20244** [comment]
This is the output required.
[out.txt](https://github.com/neams-th-coe/cardinal/files/15258873/out.txt)

**Reply by aprilnovak** [reply]
That looks good!  think there may be something wrong with just the test harness itself, but not the Cardinal install. You can run Cardinal simulations fine with the Cardinal executable, but hitting a WIN_Allocate error when running those same input files through the test harness.

If you are looking to do simulations and analysis, you should be ready to go with your Cardinal install. If you were going to be changing Cardinal's source code and needing to run the test suite frequently for test-based design, then we should still try to figure out the `run_tests`-induced failures.

**Comment by ME-20244** [comment]
As I am new in using the code, your conclusion reassures me to begin the simulation and analysis phase.
Thank you very much for your help.
